Better Configuration in TypeScript with the `satisfies` Operator/thompson.md

```diff
@@ -0,0 +1,250 @@
+예를 들어, 아래 코드에서
+
+```ts
+const routes = {
+  HOME: { path: "/" },
+} satisfies Routes;
+```
+
+만약 우리가 `path` 속성의 타입을 체크하면, 우리는 `string` 타입을 얻습니다.
+
+```ts
+routes.HOME.path; // Type: string
+```
+
+그러나 구성과 관해서는, [const 단언](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html#const-assertions) (`as const`)가 빛을 발합니다. 만약 여기서 `as const`를 사용한다면 우리는 더 정확한 문자열 리터럴 `'/'` 타입을 얻을 수 있습니다.
+
+```ts
+const routes = {
+  HOME: { path: "/" },
+} as const;
+
+routes.HOME.path; // Type: '/'
+```
+
+"그래, 이건 교묘한 수법인데 내가 왜 신경 써야돼"와 같이 생각할 수 있습니다. 그리나 아래와 같이 매개변수로 받는 정확한 경로까지 type-safe한 메소드를 고려해봅시다.(여기 문장을 어떻게 번역해야할 지 애매함)
+
+```ts
+function navigate(path: '/' | '/auth') { ... }
+```
+
+만약 우리가 `satisfies`만 사용한다면, 각 `path`가 문자열만 될 수 있다고 알려질 경우, 타입 에러가 발생하게 됩니다.
+
+```ts
+const routes = {
+  HOME: { path: "/" },
+} satisfies Routes;
+
+navigate(routes.HOME.path);
+// ❌ Argument of type 'string' is not assignable to parameter of type '"/" | "/auth"'
+```
+
+윽! `HOME.path`는 유효한 문자열이지만(`'/'`), 타입스크립트는 그렇지 않다고 말하고 있습니다.
+
+이것이 우리가 `satisfies`와 `as const`를 결합해서 가장 좋은 결과를 얻을 수 있는 부분입니다.
+
+```ts
+  const routes = {
+    HOME: { path: '/' }
+- } satisfies Routes
++ } as const satisfies Routes
+```
+
+이제, 우리는 우리가 사용하는 정확한 리터럴 값까지 타입 체크하는 좋은 해결법을 얻었습니다.
+
+```ts
+const routes = {
+  HOME: { path: "/" },
+} as const satisfies Routes;
+
+navigate(routes.HOME.path); // ✅ - as desired
+navigate("/invalid-path"); // ❌ - as desired
+```
+
+마지막으로, 왜 `satisfies` 대신에 그냥 `as const`를 사용하면 안되는건지 물어볼 수 있습니다.  
+음, `as const`만 사용하면, 우리가 그것을 작성할 때에 객체의 타입 체크를 할 수 없습니다. 그것은 우리의 IDE가 자동완성을 할 수 없거나 오타나 작성하는 동안 발생하는 다른 이슈들을 알려주지 않는다는 의미입니다.  
+이것이 `satisfies`와 `as const`를 결합하는 것이 효율적인 이유입니다.
+
+## Coming to libraries and frameworks near you (coming to 를 뭐라고 번역해야 할까요?)
+
+당신이 이 새로운 연산자로부터 얻을 수 있는 가장 큰 이점은 당신이 사용하는 유명한 라이브러리와 프레임워크에서 지원한다는 것입니다.  
+예를 들어 NextJS에서, 이전의 방식으로 작성하면
+
+```ts
+export const getStaticProps: GetStaticProps = () => {
+  return {
+    hello: 'world'
+  }
+}
+
+export default function Page(
+  props: InferGetStaticPropsType<typeof getStaticProps>
+) {
+  // 🚩 Typo not caught, because `props` is of type `{ [key: string]: any }`
+  return <div>{props.hllo></div>
+}
+```
+
+이 경우 `props`의 타입은 `{ [key: string]: any}` 입니다. 하, 도움이 되지 않네요.  
+그러나 `satisfies`와 함께라면, 당신은 이렇게 작성할 수 있습니다.
+
+```ts
+export const getStaticProps = () => {
+  return {
+    hello: 'world'
+  }
+} satisfies GetStaticProps
+
+export default function Page(
+  props: InferGetStaticPropsType<typeof getStaticProps>
+) {
+  // 😍 `props` is of type `{ hello: string }`, so our IDE makes sure our
+  // code is correct
+  return <div>{props.hello}</div>
+}
+```
+
+이렇게 작성하면 당신이 이전에 했던 타입 체크와 완성 기능을 얻을 수 있지만, 이제 적절한 타입 `{ hello: string }`이 됩니다.
+
+## 결론
+
+타입스크립트 4.9는 새로운 `satisfies` 키워드를 소개했고, 이것은 대부분의 타입스크립트의 구성 관련 작업에서 매우 편리합니다.  
+표준 타입 정의와 비교해서, 이것은 타입 체크와 최적의 타입 안전성을 위한 명확한 세부 구성 정보를 이해하는 것 사이의 우아한 균형을 이룰 수 있습니다. (여기 문장이 제일 어려운 것 같아요)  
```

Comment:
> Compared to a standard type declaration, it can strike an elegant balance between type checking and understanding the exact details of your configuration for optimal type safety and a great in-IDE.

여기 뭔가 추상적이어서 저도 많이 헷갈렸습니다. ㅎㅎㅎㅎㅎ
정기님 해석도 나쁘지 않은 것 같습니다 ㅎㅎㅎ 

다음은 제 해석인데 혹시 도움이 되실 수 있을 수 있을 것 같아서 적어놨습니다.


> 표준 type 선언을 비교할 때, 바람직한 type 안전과 내부-IDE에서 잘 동작하는 것을 위해 type 검사와 정확한 세부 정보의 이해 사이에서 균형을 이루어야 합니다.